# pRESTO Example

This follows along with the [pRESTO example] for antibody sequencing analysis
for non-UMI Illumina data, using 25,0000 2x250 nt reads of their data from
[ERR346600].  In their workflow the reverse read starts in the V region and
heads toward J, whereas the forward read starts in the constant region and
heads toward V.  The rules in `Makefile` do essentially the same thing as their
example shell script in the [example tarball].

    conda env update --file environment.yml
    conda activate example-presto
    make
    make check

## Note on Reproducibility

You won't necessarily get the exact same sequence order or representative
sequence for duplicates between runs with the same input files.  Several of the
scripts use multithreading by default, so I suspect `--nproc 1` in those cases
might make the output completely deterministic, but I haven't tested it.  The
actual sequence content is identical either way, of course.

## Example Read Pair

Following `ERR346600.14327` along the pipeline, from raw read pair to
assembled, filtered, deduplicated, and primer-assigned seqeuence:

### Paired-end Assembly

`ERR346600_1.fastq`:

    @ERR346600.14327 BS-DSFCONTROL04:4:000000000-A3F0Y:1:1101:14321:12167 length=251
    ATGCCAATGGATAAACAGATGGGGGTGTCGTTTTGGCTGAGAAGACGGTGACTGAGGTTCCTTGACCCCAGTAGTCCATAGCGCCATACCTCATTCTCGACCTCGCACAGTAATACATGGCTGTGTCCTCAGACTTCAGACTACTCATTTGCAGGTACAGGTTATTCTTGGCATTGTCTCTGGAGATGGTGAATCGCCCCTTCACACTGTCTGGATAGTAGGAGTAATTACCACCATCACTTATGGTTGCG

`ERR346600_2.fastq`:

    @ERR346600.14327 BS-DSFCONTROL04:4:000000000-A3F0Y:1:1101:14321:12167 length=251
    GTGAGACGTGATGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAA

`AP.log`:

          ID> ERR346600.14327
        SEQ1> GTGAGACGTGATGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAA
        SEQ2>                                                                                                                                                    CGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT
    ASSEMBLY> GTGAGACGTGATGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT
      LENGTH> 398
     OVERLAP> 104
       ERROR> 0.0000
      PVALUE> 1.1102e-16

Note that in this implementation, SEQ1 is the R2 read taken as-is, and SEQ2 is the R1 read reverse-complemented.

`AP_table.tab`:

    ID              LENGTH  OVERLAP ERROR   PVALUE
    ERR346600.14327 398     104     0.0000  1.1102e-16

`M1_assemble-pass.fastq`

    @ERR346600.14327
    GTGAGACGTGATGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT

### Quality Control

`M1_quality-pass.fastq`: as in `M1_assemble-pass.fastq`.

`FS.log`:

         ID> ERR346600.14327
        SEQ> GTGAGACGTGATGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT
    QUALITY> 38.432160804020

`FS_table.tab`:

    ID              QUALITY
    ERR346600.14327 38.4321608040201

#### Forward

`MPV.log`:

          ID> ERR346600.14327
      PRIMER> VH-FW9
    PRORIENT> F
     PRSTART> 4
       INSEQ> GTGAGACGTGATGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT
       ALIGN> ----GAVGTGAWGYTGGTGGAGTC
      OUTSEQ>     NNNNNNNNNNNNNNNNNNNNTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT
       ERROR> 0.0

The merged sequence is INSEQ, and V primer `VH-FW9` from
`Greiff2014_VPrimers.fasta` is found near the beginning, after four random
nucleotides.  The output trims off those four and then masks the primer
sequence.

`MPV_table.log`:

    ID              PRIMER  ERROR
    ERR346600.14327 VH-FW9  0.0

`M1-FWD_primers-pass.fastq`

    @ERR346600.14327|VPRIMER=VH-FW9
    NNNNNNNNNNNNNNNNNNNNTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT

#### Reverse

`MPC.log`:

          ID> ERR346600.14327|VPRIMER=VH-FW9
      PRIMER> IGHG
    PRORIENT> RC
     PRSTART> 370
       INSEQ> NNNNNNNNNNNNNNNNNNNNTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACACCCCCATCTGTTTATCCATTGGCAT
       ALIGN>                                                                                                                                                                                                                                                                                                                                                                                   CCCCATCDGYYYATCCMYTG----
      OUTSEQ> NNNNNNNNNNNNNNNNNNNNTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACAC
       ERROR> 0.0

Here V-primer-masked sequence is INSEQ, and C primer `IGHG` from
`Greiff2014_CPrimers.fasta` is found near the end, before four random
nucleotides.  The output trims off the entire primer and trailing four
nucleotides.

`MPC_table.log`:

    ID                              PRIMER  ERROR
    ERR346600.14327|VPRIMER=VH-FW9  IGHG    0.0

`M1-REV_primers-pass.fastq`

    @ERR346600.14327|VPRIMER=VH-FW9|CPRIMER=IGHG
    NNNNNNNNNNNNNNNNNNNNTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACAC

At this point the sequence is labeled for both primers, with the V-region
primer masked, the C-region primer trimmed, and both 4-nucleotide ends trimmed.

### Deduplication and Filtering

`M1_collapse-unique.fastq`:

    @ERR346600.14327|CPRIMER=IGHG|VPRIMER=VH-FW1,VH-FW10,VH-FW11,VH-FW12,VH-FW13,VH-FW15,VH-FW2,VH-FW3,VH-FW4,VH-FW5,VH-FW8,VH-FW9|DUPCOUNT=179
    NNNNNNNNNNNNNNNNNNNNTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATTATATATATTGGCTTCGCCAGACTCCGGAAAAGAGGCTGGAGTGGGTCGCAACCATAAGTGATGGTGGTAATTACTCCTACTATCCAGACAGTGTGAAGGGGCGATTCACCATCTCCAGAGACAATGCCAAGAATAACCTGTACCTGCAAATGAGTAGTCTGAAGTCTGAGGACACAGCCATGTATTACTGTGCGAGGTCGAGAATGAGGTATGGCGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTTCTCAGCCAAAACGACAC

Here sequences unique once primers are excluded are combined, and all the
contributing primers are noted in the sequence ID.  There seems to be an
implicit assumption that no barcoding is included in the entire process, since
the primers are all considered together and the only mention of barcoding is in
other workflows that use fixed-length UMI sequences.

`M1_atleast-2.fastq`: as in `M1_collapse-unique.fastq` since `DUPCOUNT=179`.

`M1_atleast-2_headers.tab`:

    ID              DUPCOUNT        CPRIMER VPRIMER
    ERR346600.14327 179             IGHG    VH-FW1,VH-FW10,VH-FW11,VH-FW12,VH-FW13,VH-FW15,VH-FW2,VH-FW3,VH-FW4,VH-FW5,VH-FW8,VH-FW9


[pRESTO example]: https://presto.readthedocs.io/en/stable/workflows/Greiff2014_Workflow.html
[ERR346600]: https://www.ncbi.nlm.nih.gov/sra/ERR346600
[example tarball]: http://clip.med.yale.edu/immcantation/examples/Greiff2014_Example.tar.gz
